src/Compilers/CSharp/Portable/FlowAnalysis/PreciseAbstractFlowPass.cs

```diff
@@ -310,8 +311,8 @@ protected override bool ConvertInsufficientExecutionStackExceptionToCancelledByS
         }
 
         /// <summary>
-        /// A pending branch.  There are created for a return, break, continue, goto statement,
-        /// yield return, yield break, await expression, and if PreciseAbstractFlowPass.trackExceptions
+        /// A pending branch.  These are created for a return, break, continue, goto statement,
+        /// yield return, yield break, await expression, foreach/using await, and if PreciseAbstractFlowPass.trackExceptions
```

Comment:
>using await [](start = 65, length = 11)

Could you explain how it is useful for ```using await```? What aspect would not work if we don't create the pending branch?

Reply:
Without pending branches, we'd get a warning in an `async` method despite it having an `using await`, such as:
```
  // (4,53): warning CS1998: This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread.
                //     public static async System.Threading.Tasks.Task Main()
                Diagnostic(ErrorCode.WRN_AsyncLacksAwaits, "Main").WithLocation(4, 53)
```

Reply:
> Without pending branches, we'd get a warning in an async method despite it having a using await, such as:

It feels like there is an easier, more straight forward way to detect if there is an await in a method. If that is the only benefit that we are getting from using pending branches for await constructs, then I would consider simplifying implementation.

---
In reply to: [161345316](https://github.com/dotnet/roslyn/pull/23961#discussion_r161345316) [](ancestors = 161345316)

Reply:
There is a second case that's impacted. See test below.
I think that case offers a good justification for producing a pending branch for both await expressions and async usings.

```C#
        [Fact]
        public void TestInFinallyBlock()
        {
            string source = @"
class C : System.IAsyncDisposable
{
    async System.Threading.Tasks.Task<int> M()
    {
        try
        {
        }
        finally
        {
            using await (var x = new C())
            {
            }
        }
        return 1;
    }
    public System.Threading.Tasks.Task DisposeAsync()
    {
        throw null;
    }
}
";
            var comp = CreateCompilationWithMscorlib46(source + s_interfaces, options: TestOptions.UnsafeDebugDll);
            comp.VerifyDiagnostics(
                // (11,13): error CS0157: Control cannot leave the body of a finally clause
                //             using await (var x = new C())
                Diagnostic(ErrorCode.ERR_BadFinallyLeave, "using").WithLocation(11, 13)
                );
        }
```

Reply:
I got this wrong. Awaits are indeed allowed inside finally. So should async usings.

---
In reply to: [161408348](https://github.com/dotnet/roslyn/pull/23961#discussion_r161408348) [](ancestors = 161408348)